### PR TITLE
speed up dataset concat in OCL scenarios

### PR DESCRIPTION
**Pull request by AntonioCarta**
improve the flattening process by recognizing when the same dataset occurs multiple times in the dataset list:
- added `_flatten_and_reindex` to improve flattening
- added a `__hash__` method for `AvalancheDataset` to recognize multiple dataset occurrences

@lrzpellegrini can you check this?

**Update**: fixed it. Now it's fast for the entire stream:
```
(split MNIST OCL with exp-size=5 samples, 6000 iterations total)
6003it [00:15, 389.10it/s]
ReservoirSampling Duration:  15.429586410522461
6003it [00:29, 205.35it/s]
ParametricBuffer (random sampling) Duration:  29.233972787857056
```

### CODE REVIEW

File: tests/benchmarks/test_flat_data.py [hunk 1]
```diff
@@ -88,3 +96,127 @@ def test_merging(self):
...
+        print(f"DATA depth={_flatdata_depth(B)}, dsets={len(B._datasets)}")
+        assert _flatdata_depth(B) == 2
+        assert len(B._datasets) == 1
+
+        B = D.concat(B)
+        print(f"DATA depth={_flatdata_depth(B)}, dsets={len(B._datasets)}")
+        assert _flatdata_depth(B) == 2
+        assert len(B._datasets) == 1
+
+    def test_concat_flattens_nc_scenario_dataset(self):
```
Comment (lrzpellegrini) [inline tests/benchmarks/test_flat_data.py:150]: There are no asserts here

File: avalanche/benchmarks/utils/data.py [hunk 2]
```diff
@@ -194,13 +194,17 @@ def __init__(
...
                 # TODO: this was the old behavior. How do we know what to do if
                 # we permute the entire dataset?
-                if len(da) != sum([len(d) for d in datasets]):
-                    self._data_attributes[da.name] = da
-                else:
-                    self._data_attributes[da.name] = da.subset(self._indices)
-
-                    dasub = da.subset(indices)
-                    self._data_attributes[da.name] = dasub
+                # DEPRECATED! always subset attributes
```
Comment (lrzpellegrini) [inline avalanche/benchmarks/utils/data.py:197]: It may make sense to add a unit test to cover this change.

**Comment by lrzpellegrini** [comment]
The PR makes sense to me. Correct me if I'm wrong, but will it be able to flatten most replay buffers into a single dataset (apart from benchmarks created with dataset_benchmarks)? 

I only left a couple of comments, everything else seems ok to me!

**Comment by AntonioCarta** [comment]
> Correct me if I'm wrong, but will it be able to flatten most replay buffers into a single dataset (apart from benchmarks created with dataset_benchmarks)?

yes! The details depend on the benchmark creation. For example, Permuted MNIST experiences are considered separate datasets because of the transformations.

**Comment by AntonioCarta** [comment]
Thanks @lrzpellegrini I added the tests and an error to catch the backward incompatible behavior.

**Pull request merged**